src/hdmf/container.py

```diff
@@ -749,10 +755,71 @@ def _generate_field_html(self, key, value, level, access_code):
 
         return html_repr
 
+
     def _generate_array_html(self, array, level):
-        """Generates HTML for a NumPy array."""
-        str_ = str(array).replace("\n", "</br>")
-        return f'<div style="margin-left: {level * 20}px;" class="container-fields">{str_}</div>'
+        """Generates HTML for a NumPy array, h5py Dataset, or Zarr array."""
+
+        def convert_bytes_to_str(bytes_size):
+            suffixes = ['bytes', 'KiB', 'MiB', 'GiB', 'TiB', 'PiB']
+            i = 0
+            while bytes_size >= 1024 and i < len(suffixes)-1:
+                bytes_size /= 1024.
+                i += 1
+            return f"{bytes_size:.2f} {suffixes[i]}"
+
+        # Generates an html report for the backend info, inspired on zarr info html representation
+        def html_table(item_dicts) -> str:
+            report = '<table class="data-info">'
+            report += "<tbody>"
+            for k, v in item_dicts.items():
+                report += (
+                    f"<tr>"
+                    f'<th style="text-align: left">{k}</th>'
+                    f'<td style="text-align: left">{v}</td>'
+                    f"</tr>"
+                )
+            report += "</tbody>"
+            report += "</table>"
+            return report
+
+        array_size_in_bytes = array.nbytes
+        array_size_repr = convert_bytes_to_str(array_size_in_bytes)
+        basic_array_info_dict = {"shape": array.shape, "dtype": array.dtype,  "Array size": array_size_repr}
+
+        if isinstance(array, np.ndarray):
+            head = "NumPy Array"
+            backend_info_dict = basic_array_info_dict
+
+        if isinstance(array, h5py.Dataset):
+            hdf5_dataset = array
+            chunks = hdf5_dataset.chunks
+            compression = hdf5_dataset.compression
+            uncompressed_size = hdf5_dataset.nbytes
+            compression_opts = hdf5_dataset.compression_opts
+            compressed_size = hdf5_dataset.id.get_storage_size()
+            compression_ratio = uncompressed_size / compressed_size if compressed_size != 0 else "undefined"
+
+            head = "HDF5 Dataset"
+            hdf5_info_dict = {"chunks": chunks, "compression": compression, "compression_opts": compression_opts,
+                              "compression_ratio": compression_ratio}
+            backend_info_dict = {**basic_array_info_dict, **hdf5_info_dict}
+
+        if hasattr(array, "store") and hasattr(array, "shape"):  # Duck typing for zarr array
+            head = "Zarr Array"
+            zarr_info_dict = {k:v for k, v in array.info_items()}
+            backend_info_dict = zarr_info_dict
```

Comment:
It would be nice to avoid having logic that is specific for a particular io backend in the `Container`. The reason is that this inhibts implementing backends in a self-contained manner and stand-alone packages and requires updating many places throughout HDMF. 

The checks for `h5py.Dataset` and `Zarr.array` are really only relevant when a file was read from file. To help disentangle the dependencies, I'm wondering whether we could do the following:

1. Add a static method ``generate_dataset_html`` to ``HDMFIO`` that would then need to implemented by ``HDF5IO`` and ``ZarrIO``
2. In the ``Container`` you could then do something like:

```
read_io = self.get_read_io()  # if the Container was read from file, this will give you the IO object that read it
if read_io is not None:
    html_repr = read_io.generate_dataset_html(my_dataset)
else:
    # The file was not read from disk so the dataset should be numpy array or a list
```



`

Reply:
> It would be nice to avoid having logic that is specific for a particular io backend in the Container. The reason is that this inhibts implementing backends in a self-contained manner and stand-alone packages and requires updating many places throughout HDMF.

I see, yes, it would be nice if we could decouple this.  On the other hand, right now, if they do implement their own backend they will just lose the representation for datasets which is not critical. 

> The checks for h5py.Dataset and Zarr.array are really only relevant when a file was read from file.

Is it? Right now in the test we are passsing an hdf5 dataset as data without writing the nwbfile for testin the display. Is this not intended?

This proposal seems very good:
```
read_io = self.get_read_io()  # if the Container was read from file, this will give you the IO object that read it
if read_io is not None:
    html_repr = read_io.generate_dataset_html(my_dataset)
else:
    # The file was not read from disk so the dataset should be numpy array or a list
```

I see two downsides:
1) Missing extensive representation for in-memory files (it is nice to know what you will write!).
2) Fragmenting the code base for html representations.

Is there any other backend in the works right now? If not, maybe we can do this simpler way and add the complexity once we are closer to need it?

Reply:
I think it would worth giving this a try here. If it works for HDF5, we can then we can easily move the logic for Zarr to hdmf-zarr. I don't think it should be too hard to make this work right now, but these things tend to get hard to change later on.

Reply:
What about handling in-memory objects as well

Reply:
In-memory-only objects (i.e., numpy arrays and lists) can be handled here.

Reply:
But you can have in-memory h5py datasets and zarr arrays, right? Am I wrong about this? I pointed out above that that is how we are testing this.

Ok, I think that the easiest thing is to restrict this PR to just displaying information that is encoded in the array standard:

https://data-apis.org/array-api/latest/API_specification/array_object.html#attributes

So no backend specific information like compression options, chunks, etc here. We can then follow your approach in another PR to add backend related information extracted through the `DataIO` objects.

What do you think?

Reply:
> But you can have in-memory h5py datasets and zarr arrays, right?

As far as I know, h5py datasets are always tied to a file on disk (and I believe in Zarr as well). There are ways (at least in Zarr) to have the whole file then live in memory only, but the datasets are still tied to a file that is being read from. 

> So no backend specific information like compression options, chunks, etc here. We can then follow your approach in another PR to add backend related information extracted through the `DataIO` objects.

Sure, that sounds good to me.

Reply:
> As far as I know, h5py datasets are always tied to a file on disk (and I believe in Zarr as well). There are ways (at least in Zarr) to have the whole file then live in memory only, but the datasets are still tied to a file that is being read from.

What I was asking is that if for those examples (like the one in the test here) the trick of using the `DataIO` would work. I think not but maybe I am missing something.

Reply:
> We can then follow your approach in another PR to add backend related information extracted through the `DataIO` objects.

Could you point me to the PR you are referring to. I'm not sure what role `DataIO` plays for this PR.

Reply:
Yes that looks good. Just to avoid confusion, `read_io` is an instance of `HDMFIO` (i.e., `HDF5IO` or `ZarrIO`) and not `DataIO`. To implement the logic we would then need to:

1. Add `HDMFIO.generate_dataset_html(dataset)` which would implement just a minimalist representation
2. Implement `HDF5IO.generate_dataset_html(h5py.Dataset)` to represent `h5py.Dataset`
3. In a separate PR on `hdmf_zarr` implement `ZarrIO.generate_dataset_html(Zarr.array)`

To simplify this implementation and generate consistent representations, we could make a utility function that would take information about a dataset (e.g,. name, shape, data type, etc.) as input and generate the html representation such that the individual `generate_data_html` on the I/O backends would just collect the information from the dataset and use the utility function to generate the actual html.

Reply:
> Yes that looks good 👍 . Just to avoid confusion, read_io is an instance of HDMFIO (i.e., HDF5IO or ZarrIO) and not DataIO. To implement the logic we would then need to.

Yes, I realized after that I was confusing these two objects.

Reply:
Sounds good. Thanks for your hard work on this PR and the fruitful discussion.